File: 8_last_digit_of_the_sum_of_squares_of_fibonacci_numbers/fibonacci_sum_squares.py
"""
Author: Jehadur Rahman
Organization: CyArm
GitHub: https://github.com/JehadurRE
Website: jehadurre.me
Date: 2025-07-10 09:43 (Thu)
"""

def fibonacci_sum_squares_naive(n):
    if n <= 1:
        return n

    previous, current, sum_val = 0, 1, 1

    for _ in range(n - 1):
        previous, current = current, previous + current
        sum_val += current * current

    return sum_val % 10

def fibonacci_sum_squares_fast(n):
    n = n % 60
    if n <= 1:
        return n
    
    previous = 0
    current = 1
    sum_val = 1
    
    for i in range(n - 1):
        tmp_previous = previous
        previous = current
        current = (tmp_previous + current) % 10
        sum_val += current * current
    
    return sum_val % 10

if __name__ == '__main__':
    n = int(input())
    print(fibonacci_sum_squares_fast(n))

File: 8_last_digit_of_the_sum_of_squares_of_fibonacci_numbers/FibonacciSumSquares.java
/*
 * Author: Jehadur Rahman
 * Organization: CyArm
 * GitHub: https://github.com/JehadurRE
 * Website: jehadurre.me
 * Date: 2025-07-10 09:43 (Thu)
 */

import java.util.*;

public class FibonacciSumSquares {
    private static long getFibonacciSumSquaresNaive(long n) {
        if (n <= 1)
            return n;

        long previous = 0;
        long current = 1;
        long sum = 1;

        for (long i = 0; i < n - 1; ++i) {
            long tmp_previous = previous;
            previous = current;
            current = tmp_previous + current;
            sum += current * current;
        }

        return sum % 10;
    }

    private static int fibonacciSumSquaresFast(long n) {
        n = n % 60;
        if (n <= 1)
            return (int) n;

        long previous = 0;
        long current = 1;
        long sum = 1;

        for (long i = 0; i < n - 1; ++i) {
            long tmp_previous = previous;
            previous = current;
            current = (tmp_previous + current) % 10;
            sum += current * current;
        }

        return (int) (sum % 10);
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        long n = scanner.nextLong();
        // long s = getFibonacciSumSquaresNaive(n);
        int s = fibonacciSumSquaresFast(n);
        System.out.println(s);
    }
}

